store/src/java/com/zimbra/cs/service/mail/ToXML.java

```diff
@@ -2967,6 +3024,143 @@ private static void addContent(Element elt, MPartInfo mpi, int maxSize, boolean
+            }
+        } else if (ctype.equals(MimeConstants.CT_TEXT_ENRICHED)) {
+            // Enriched text handling is a little funky because TextEnrichedHandler
+            // doesn't use Reader and Writer.  As a result, we truncate
+            // the source before converting to HTML.
+            InputStream stream = mp.getInputStream();
+            Reader reader = null;
+            try {
+                reader = Mime.getTextReader(stream, mp.getContentType(), defaultCharset);
+                int maxChars = (maxSize > 0 ? maxSize + 1 : -1);
+                String enriched = ByteUtil.getContent(reader, maxChars, false);
+                if (enriched.length() == maxChars) {
+                    // The normal check for truncated data won't work because
+                    // the length of the converted text is different than the length
+                    // of the source, so set the attr here.
+                    wasTruncated = true;
+                }
+                if (wantContent.equals(MsgContent.original) || wantContent.equals(MsgContent.both)) {
+                    originalContent = TextEnrichedHandler.convertToHTML(removeQuotedText(enriched,
+                            false));
+                }
+                data = TextEnrichedHandler.convertToHTML(enriched);
+            } finally {
+                ByteUtil.closeStream(stream);
+                Closeables.closeQuietly(reader);
+            }
+        } else {
+            InputStream stream = mp.getInputStream();
+            Reader reader = null;
+            try {
+                reader = Mime.getTextReader(stream, mp.getContentType(), defaultCharset);
+                int maxChars = (maxSize > 0 ? maxSize + 1 : -1);
+                data = ByteUtil.getContent(reader, maxChars, false);
+                if (wantContent.equals(MsgContent.original) || wantContent.equals(MsgContent.both)) {
+                    originalContent = removeQuotedText(data, false);
+                }
+                if (data.length() == maxChars) {
+                    wasTruncated = true;
+                }
+            } finally {
+                ByteUtil.closeStream(stream);
+                Closeables.closeQuietly(reader);
+            }
+        }
+
+        if (data != null && !wantContent.equals(MsgContent.original)) {
+            data = StringUtil.stripControlCharacters(data);
+            if (wasTruncated) {
+                elt.addAttribute(MailConstants.A_TRUNCATED_CONTENT, true);
+                if (data.length() > maxSize) {
+                    data = data.substring(0, maxSize);
+                }
+            }
+            if (!data.contains("<html>")) {
+                String bodyContent = data;
+                String htmBodyWithWarning = "<html><body><div><p>"+warningMsg+"</p><hr /></div><div style=\"font-size: 12pt; font-family: arial, helvetica, sans-serif; color: rgb(0, 0, 0);\"><div>&nbsp;</div><div>"+bodyContent+"</div>\n";
```

Comment:
bodyContent may have <html> tag already, so are we duplicating it?

Reply:
This is a check to see if the html content is already present. If the tag is not present, we are creating the html content(warning + body).